### PR TITLE
Add RTK support via rtcm

### PR DESCRIPTION
**Pull request by r0gi**
These changes are based on [this outdated repo](https://github.com/ros-agriculture/ublox_f9p) and add RTK support for high precision GNSS modules like the ZED-F9P.  The RTK support is based on rtcm corrections that are published via ROS and this creates a dependency to [rtcm_msgs](https://github.com/tilk/rtcm_msgs). The rtcm message itself is very basic (`Header + uint8 []`). 

I've tested this with my ZED-F9P and it works reliably. The changes would address #119,  #117, and #43.

All you would need to get RTK working now is to use an NTRIP client (or similar), that publishes RTCM messages via ROS on the `/rtcm` topic.

### CODE REVIEW

**Comment by versatran01** [comment]
Thanks for the PR. 
I don't have a unit to test with but this seems like a small and straightforward change.
The only question I have is that this adds a dependency to `rtcm_msgs`.
Since this msg is very simple, can we just add it directly to this repo (with a link to the original one in the comment)?

**Comment by r0gi** [comment]
I was considering that but in that case, the ROS publisher of the RTCM message would have to add ublox_msgs as a dependency. If every GPS package had their own version of this message, it wouldn't be very modular as you'd have to change the message type for each different gps unit. 

Because the message is so simple, perhaps just using [std_msgs/UInt8MultiArray](https://docs.ros.org/en/api/std_msgs/html/msg/UInt8MultiArray.html) would be more appropriate?

I will also add an update to the readme regarding the RTK support since it seems to be a common question on how to set it up.

**Comment by Timple** [comment]
I think the `rtcm_msgs` solution is much nicer. `std_msgs` are actually deprecated because they are 'meaningless'.

Now you can look for packages which all use `rtcm_msgs` and have them interact. Everyone in the ROS ecosystem should know how to install dependencies I guess.

(`rtcm_msgs` should be listed in the CMakeListst.txt and package.xml by the way)

**Comment by versatran01** [comment]
Is there a way to make this dependency optional?

**Comment by Timple** [comment]
Having the dependency released might reduce the need for that?
https://github.com/tilk/rtcm_msgs/issues/1

**Comment by versatran01** [comment]
Sure that sounds good. @Timple i will leave this to you. Feel free to merge it whenever you see fit.

**Comment by Timple** [comment]
> (rtcm_msgs should be listed in the CMakeListst.txt and package.xml by the way)

@r0gi would you mind picking this up?

**Comment by flynneva** [comment]
>> (rtcm_msgs should be listed in the CMakeListst.txt and package.xml by the way)
>@r0gi would you mind picking this up?

isnt `rtcm_msgs` already listed in the CMakeList and package.xml files? it looks like they're added in there in this PR...or were you talking about somewhere else?

**Comment by r0gi** [comment]
I did list the dependency in the `package.xml`, but I actually forgot to put it in the `CMakelists.txt` file because `catkin_make` actually built everything successfully without having to list `rtcm_msgs` in either `package.xml` or `CMakelists.txt`. It  should be listed in both files so I'll update that.

**Comment by Timple** [comment]
Yup! `catkin_make` is much too forgiving in that. That will lead to issues later. I typically use `catkin_tools` for more isolated builds (and more beautiful output 😄).

Tip: `catkin_lint` also is a useful tool to detect these kind of problems.

You're almost there, it should also be listed in the CMakeLists.txt under `catkin_package`/`CATKIN_DEPENDS`.

**Comment by Timple** [comment]
Waiting for release of rtcm_msgs:
https://github.com/ros/rosdistro/pull/30334
https://github.com/ros/rosdistro/pull/30335

I propose to merge this only after a sync for both melodic as noetic has been made. So that people who build `ublox_gps` from source can use rosdep/apt.

**Review by Timple — APPROVED** [review]

**Comment by versatran01** [comment]
@Timple This seems to be merged into rosdistro? Unless we are still waiting for a sync.

**Comment by Timple** [comment]
Melodic sync is done, Noetic not yet.

I'm fine with merging anyhow by the way.

**Comment by arpaterson** [comment]
This addresses externally sourced RTCM corrections to be consumed by a rover. 
What about RTCM messages produced by the device as a stationary/fixed base? Is it possible for this node to publish RTCM messages to the /rtcm topic?

I'd like to use (without internet access) two F9Ps on two (or more) ros masters - one with known or surveyed-in location, producing rtcm data, bridged via mqtt to the rover(s).

**Comment by yulan0215** [comment]
Hi:

I am sorry but I have a question about the Ntrip client, I did not get it why when I ran the roslaunch to run the Ntrip CLIENT but there was not RTCM data come. Even when I deplugged the F9P USB port. The ROS ntrip still running but no data come, do you know this king of problem or if I should config some parameter before?

**Comment by rahdal** [comment]
> This addresses externally sourced RTCM corrections to be consumed by a rover. What about RTCM messages produced by the device as a stationary/fixed base? Is it possible for this node to publish RTCM messages to the /rtcm topic?
> 
> I'd like to use (without internet access) two F9Ps on two (or more) ros masters - one with known or surveyed-in location, producing rtcm data, bridged via mqtt to the rover(s).

Has a solution been found to this?

**Comment by bryangd34** [comment]
@rahdal 
Any solutions been found? Thanks.

**Pull request merged**
